File: 050/main.cpp
#include <iostream>

using namespace std;

class Solution {
public:
    double myPow(double x, int n) {
        if (x == 1 || n == 0) return 1;
        if (n == 1) return x;
        // use -(n + 1) to avoid MIN_VALUE case
        if (n < 0) return 1.0 / (x * myPow(x, -(n + 1)));
        double res = 1;
        double num = x;
        while (n > 1) {
            if (n % 2 == 1) {
                res *= num;
                n--;
            } else {
                num *= num;
                n /= 2;
            }
        }
        return res * num;
    }
};

int main()
{
    Solution sol;
    double res = sol.myPow(2.0000, -1);
    cout << res << endl;
    return 0;
}

File: 050/main.js
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function (x, n) {
  if (x === 1 || n === 0) return 1
  if (n === 1) return x
  if (n < 0) return 1 / myPow(x, -n)
  let res = 1
  let num = x
  while (n > 1) {
    if (n % 2 === 1) {
      res *= num
      n--
    } else {
      num *= num
      n /= 2
    }
  }
  return res * num
}

// console.log(myPow(2.00000, 10))
// console.log(myPow(34.00515, -3))
console.log(myPow(8.95371, -1))
